tests/interfaces/test_batch_torch.py

```diff
@@ -0,0 +1,947 @@
+                interface="torch",
+            )[0]
+
+        # Without caching, 3 evaluations are required.
+        # 1 for the forward pass, and one per output dimension
+        # on the backward pass.
+        torch.autograd.functional.jacobian(lambda x: cost(x, cache=None), params)
+        assert dev.num_executions == 3
+
+        # With caching, only 2 evaluations are required. One
+        # for the forward pass, and one for the backward pass.
+        dev._num_executions = 0
+        torch.autograd.functional.jacobian(lambda x: cost(x, cache=True), params)
+        assert dev.num_executions == 2
+
+
+torch_devices = [None]
+
+if torch.cuda.is_available():
+    torch_devices.append(torch.device("cuda"))
+
+
+execute_kwargs = [
+    {"gradient_fn": param_shift, "interface": "torch"},
+    {
+        "gradient_fn": "device",
+        "mode": "forward",
+        "gradient_kwargs": {"method": "adjoint_jacobian", "use_device_state": True},
+        "interface": "torch",
+    },
+    {
+        "gradient_fn": "device",
+        "mode": "backward",
+        "gradient_kwargs": {"method": "adjoint_jacobian"},
+        "interface": "torch",
+    },
+]
+
+
+@pytest.mark.parametrize("torch_device", torch_devices)
+@pytest.mark.parametrize("execute_kwargs", execute_kwargs)
+class TestTorchExecuteIntegration:
+    """Test the torch interface execute function
+    integrates well for both forward and backward execution"""
+
+    def test_execution(self, torch_device, execute_kwargs):
+        """Test execution"""
+        dev = qml.device("default.qubit", wires=1)
+        a = torch.tensor(0.1, requires_grad=True, device=torch_device)
+        b = torch.tensor(0.2, requires_grad=False, device=torch_device)
+
+        with qml.tape.JacobianTape() as tape1:
+            qml.RY(a, wires=0)
+            qml.RX(b, wires=0)
+            qml.expval(qml.PauliZ(0))
+
+        with qml.tape.JacobianTape() as tape2:
+            qml.RY(a, wires=0)
+            qml.RX(b, wires=0)
+            qml.expval(qml.PauliZ(0))
+
+        res = execute([tape1, tape2], dev, **execute_kwargs)
+
+        assert len(res) == 2
+        assert res[0].shape == (1,)
+        assert res[1].shape == (1,)
+
+    def test_scalar_jacobian(self, torch_device, execute_kwargs, tol):
+        """Test scalar jacobian calculation"""
+        a = torch.tensor(0.1, requires_grad=True, dtype=torch.float64, device=torch_device)
+        dev = qml.device("default.qubit", wires=2)
+
+        with qml.tape.JacobianTape() as tape:
+            qml.RY(a, wires=0)
+            qml.expval(qml.PauliZ(0))
+
+        res = execute([tape], dev, **execute_kwargs)[0]
+        res.backward()
+
+        # compare to standard tape jacobian
+        with qml.tape.JacobianTape() as tape:
+            qml.RY(a, wires=0)
+            qml.expval(qml.PauliZ(0))
+
+        tape.trainable_params = {0}
+        tapes, fn = param_shift(tape)
+        expected = fn(dev.batch_execute(tapes))

```

Comment:
Hmm, wouldn't we have liked to use `tape.jacobian` here? I'd imagine using `execute` (1. part of the test) to use the same `param_shift` & `dev.batch_execute` pipeline in the background, is that not the case?

Reply:
I am avoiding using `tape.jacobian` because I will be removing it in an upcoming PR 😆

Reply:
The `pennylane.gradients` functions are the 'replacements' for `tape.jacobian`

Reply:
Oh, that makes sense! In that case, how about either using the analytic expression here or checking against the output of the backprop device?

Reply:
done!